### PR TITLE
Allow multiple paths in diff path limiter

### LINKED ISSUE #584: Can't use diff with multiple paths
### Subject of the issue

You can't use the `diff` method with multiple paths.

### Your environment
• `git version 2.32.0 (Apple Git-132)`
• `git (1.10.0)`
• `ruby 3.0.3p157 (2021-11-24 revision 3fb7d2cadc) [arm64-darwin21]`

### Steps to reproduce

```ruby
puts(Git.open('.').diff.path('dir_a dir_b'))
```

### Expected behaviour

The diff for the files under the directories `dir_a` and `dir_b` should be displayed.

### Actual behaviour

No diff is displayed.

### PR DESCRIPTION
**Pull request by jcouball**
## Summary

Fixes #584

The `diff` method now accepts multiple paths via splatted arguments, allowing users to filter diffs by multiple paths at once.

## Changes

- `Git::Diff#path` now accepts varargs (`*paths`) and normalizes to string or array
- `Git::Lib` diff methods (`diff_full`, `diff_stats`, `diff_path_status`) now accept String or Array path limiters and splat them after `--`
- `Git::Base#diff_stats` and `#diff_path_status` accept an opts hash with path limiter option
- Added YARD documentation for the new `path` method signature
- Updated README with multi-path diff example
- Added tests for multi-path usage across diff, stats, and path status

## Usage

```ruby
# Single path (existing behavior)
git.diff('v1', 'v2').path('lib/')

# Multiple paths (new)
git.diff('v1', 'v2').path('lib/', 'docs/', 'README.md')

# Via diff_stats with opts
git.diff_stats('v1', 'v2', path_limiter: ['lib/', 'docs/'])
```

## Testing

All 453 tests pass. Added 3 new tests covering multi-path behavior.

### CODE REVIEW

File: lib/git/base.rb [hunk 1]
```diff
@@ -833,18 +833,22 @@ def merge_base(*)
...
     # @param objectish [String] The first commit or object to compare. Defaults to 'HEAD'.
     # @param obj2 [String, nil] The second commit or object to compare.
-    # @return [Git::Diff::PathStatus]
-    def diff_path_status(objectish = 'HEAD', obj2 = nil)
-      Git::DiffPathStatus.new(self, objectish, obj2)
+    # @param opts [Hash] Options to filter the diff.
+    # @option opts [String, Array<String>] :path Limit status to specified path(s).
+    # @return [Git::DiffPathStatus]
+    def diff_path_status(objectish = 'HEAD', obj2 = nil, opts = {})
+      Git::DiffPathStatus.new(self, objectish, obj2, opts[:path])
```
[the remark this answers is not in the document]
Comment (jcouball) [reply lib/git/base.rb]: Changed it so that all three diff-related methods (diff_full, diff_stats, diff_path_status) now consistently use :path_limiter as the option key in the internal Git::Lib API.

File: lib/git/diff.rb [hunk 2]
```diff
@@ -18,8 +18,28 @@ def initialize(base, from = nil, to = nil)
...
-      @path = path
+    # Limits the diff to the specified path(s)
+    #
+    # @example Limit diff to a single path
+    #   git.diff('HEAD~3', 'HEAD').path('lib/')
+    #
+    # @example Limit diff to multiple paths
+    #   git.diff('HEAD~3', 'HEAD').path('src/', 'docs/', 'README.md')
+    #
+    # @param paths [Array<String, Pathname>] one or more paths to filter the diff
```
[the remark this answers is not in the document]
Comment (jcouball) [reply lib/git/diff.rb]: Changed the signature.

File: tests/units/test_diff.rb [hunk 3]
```diff
@@ -48,6 +48,25 @@ def test_diff_path
...
+    assert_equal(10, d.deletions)
+    assert_equal(64, d.insertions)
+  end
+
+  def test_diff_path_empty_array
+    @git.lib.expects(:diff_full).with('gitsearch1', 'v2.5', { path_limiter: nil }).returns('')
+
+    d = @git.diff('gitsearch1', 'v2.5').path([])
+    d.patch
+  end
```
[the remark this answers is not in the document]
Comment (jcouball) [reply tests/units/test_diff.rb:68]: ## Re: test_diff_path_empty_array feedback

The test is functioning correctly as written. The `expects` call creates a strict expectation that verifies:

1. **The empty array is converted to nil** - If `path([])` didn't convert the empty array to `nil`, the expectation would fail because `diff_full` would be called with `{ path_limiter: [] }` instead of `{ path_limiter: nil }`
2. **The method is called with exact arguments** - Any deviation from the expected arguments causes the test to fail

This pattern is consistent across all three empty-array tests in this PR:
- `test_diff_path_empty_array` (test_diff.rb)
- `test_diff_stats_with_empty_path_array` (test_diff_stats.rb)  
- `test_path_status_with_empty_path_array` (test_diff_path_status.rb)

All three use `expects` to verify that empty arrays are properly normalized. This is a standard testing approach in this codebase - the mock expectation itself serves as the assertion that validates the behavior.

The test successfully verifies that the implementation correctly handles the edge case of empty input.

**Comment by jcouball** [comment]
## Update: Validation for nested arrays added

Based on the feedback about `flatten`, I've updated the implementation to **reject nested arrays with a clear error message** rather than silently flattening them.

### Changes:

**Before:**
```ruby
def path(*paths)
  flattened_paths = paths.flatten.compact
  # ... assign to @path
end
```

**After:**
```ruby
def path(*paths)
  validate_paths_not_arrays(paths)  # Raises ArgumentError if nested arrays found
  cleaned_paths = paths.compact
  # ... assign to @path
end
```

### Rationale:

Without validation, `path([['lib/', 'docs/']])` would be converted to `"[\"lib/\", \"docs/\"]"` (a string representation of an array) and passed to git as a malformed path argument, causing silent failures.

By explicitly rejecting nested arrays with a clear error message, we:
1. **Fail fast** - Catch mistakes at the API boundary
2. **Provide clear guidance** - Error message shows the correct usage
3. **Maintain API contract** - The signature `path(*paths)` expects individual arguments

### Error message:
```
ArgumentError: path expects individual arguments, not arrays. Use path('lib/', 'docs/') not path(['lib/', 'docs/'])
```

### Test coverage:
Added `test_diff_path_rejects_nested_arrays` to verify the validation works correctly.

All 459 tests pass ✓  
Rubocop clean ✓

File: lib/git/diff.rb [hunk 4]
```diff
@@ -18,8 +18,28 @@ def initialize(base, from = nil, to = nil)
...
+    #   git.diff('HEAD~3', 'HEAD').path('lib/')
+    #
+    # @example Limit diff to multiple paths
+    #   git.diff('HEAD~3', 'HEAD').path('src/', 'docs/', 'README.md')
+    #
+    # @param paths [String, Pathname, Array<String, Pathname>] one or more paths to filter the diff
+    # @return [self] returns self for method chaining
+    #
+    def path(*paths)
+      flattened_paths = paths.flatten.compact
```
[the remark this answers is not in the document]
Comment (jcouball) [reply lib/git/diff.rb]: ## Update: Validation for nested arrays added

Based on the feedback about `flatten`, I've updated the implementation to **reject nested arrays with a clear error message** rather than silently flattening them.

### Changes:

**Before:**
```ruby
def path(*paths)
  flattened_paths = paths.flatten.compact
  # ... assign to @path
end
```

**After:**
```ruby
def path(*paths)
  validate_paths_not_arrays(paths)  # Raises ArgumentError if nested arrays found
  cleaned_paths = paths.compact
  # ... assign to @path
end
```

### Rationale:

Without validation, `path([['lib/', 'docs/']])` would be converted to `"[\"lib/\", \"docs/\"]"` (a string representation of an array) and passed to git as a malformed path argument, causing silent failures.

By explicitly rejecting nested arrays with a clear error message, we:
1. **Fail fast** - Catch mistakes at the API boundary
2. **Provide clear guidance** - Error message shows the correct usage
3. **Maintain API contract** - The signature `path(*paths)` expects individual arguments

### Error message:
```
ArgumentError: path expects individual arguments, not arrays. Use path('lib/', 'docs/') not path(['lib/', 'docs/'])
```

### Test coverage:
Added `test_diff_path_rejects_nested_arrays` to verify the validation works correctly.

All 459 tests pass ✓  
Rubocop clean ✓

File: lib/git/lib.rb [hunk 5]
```diff
@@ -866,7 +890,7 @@ def diff_stats(obj1 = 'HEAD', obj2 = nil, opts = {})
 
     DIFF_PATH_STATUS_OPTION_MAP = [
       { type: :static, flag: '--name-status' },
-      { keys: [:path], type: :validate_only }
+      { keys: [:path_limiter], type: :validate_only }
     ].freeze
```
[the remark this answers is not in the document]
Comment (jcouball) [reply lib/git/lib.rb:920]: ## Update: Added deprecation handling at Git::Lib level

Good catch on the breaking change at the `Git::Lib` level. I've added consistent deprecation handling.

### Changes:

1. **Re-added `:path` to `DIFF_PATH_STATUS_OPTION_MAP`** - Both `:path` and `:path_limiter` are now accepted
2. **Added deprecation in `Git::Lib#diff_path_status`** - Issues the same deprecation warning as `Git::Base#diff_path_status`
3. **Extracted helper method** - Created `handle_deprecated_path_option` to keep method length under rubocop limit
4. **Added test coverage** - New `test_lib_path_status_path_option_deprecated` verifies the Lib-level deprecation

### Why this matters:

While `Git::Lib` is marked `@api private`, maintaining consistency in deprecation handling is important:
- **No breaking changes** - Code using `git.lib.diff_path_status(ref1, ref2, path: 'some/path')` continues to work
- **Clear migration path** - Users get deprecation warnings at both API levels
- **Consistent behavior** - Both `Base` and `Lib` handle the transition the same way

All 460 tests pass ✓  
Rubocop clean ✓

File: lib/git/diff.rb [hunk 6]
```diff
@@ -18,8 +18,28 @@ def initialize(base, from = nil, to = nil)
...
+      @path = if flattened_paths.empty?
+                nil
+              elsif flattened_paths.length == 1
+                flattened_paths.first
+              else
+                flattened_paths
+              end
+
       self
     end
```
[the remark this answers is not in the document]
Comment (jcouball) [reply lib/git/diff.rb:54]: ## Update: Enhanced YARD documentation for Git::Diff#path

The feedback about documenting the behavior is valid. I've enhanced the YARD documentation to be more comprehensive.

### Documentation improvements:

1. **Added description of internal behavior** - Explains that single paths are stored as String, multiple as Array
2. **Clarified no-argument behavior** - Documents that calling `path()` or `path(nil)` removes filtering
3. **Added example for removing filters** - Shows how to reset path filtering
4. **Updated @param description** - Clarifies that passing no arguments removes filtering

### Note on the feedback:

The feedback referenced code that used `flatten`, which we've since replaced with validation that **rejects arrays**. The current implementation:
- ✅ Does NOT accept `path(['lib/'])` - raises ArgumentError
- ✅ Only accepts `path('lib/')` or `path('lib/', 'docs/')`
- ✅ Calling `path()` with no arguments removes filtering (sets @path to nil)

The enhanced documentation now clearly explains these behaviors.

All 460 tests pass ✓  
Rubocop clean ✓

File: lib/git/lib.rb [hunk 7]
```diff
@@ -825,6 +825,30 @@ def assert_args_are_not_options(arg_name, *args)
...
+      normalized = pathspecs.is_a?(Array) ? pathspecs : [pathspecs]
+      unless normalized.all? { |path| path.respond_to?(:to_s) }
+        raise ArgumentError, "Invalid #{arg_name}: must be a String or Array of Strings"
+      end
+
+      normalized = normalized.map(&:to_s)
+      return nil if normalized.empty?
+
+      normalized
+    end
```
[the remark this answers is not in the document]
Comment (jcouball) [reply lib/git/lib.rb:851]: I have implemented validation that pathspecs is nil, String, Pathname, or an Array of Strings/Pathnames.

    # @param pathspecs [String, Pathname, Array<String, Pathname>, nil] path(s) to normalize

Anything that does not conform to this will raise an ArgumentError.

I also made sure that this method is comprehensively tested.

File: lib/git/base.rb [hunk 8]
```diff
@@ -833,18 +833,32 @@ def merge_base(*)
...
     # Returns a Git::Diff::PathStatus object for accessing the name-status report.
     #
     # @param objectish [String] The first commit or object to compare. Defaults to 'HEAD'.
     # @param obj2 [String, nil] The second commit or object to compare.
-    # @return [Git::Diff::PathStatus]
-    def diff_path_status(objectish = 'HEAD', obj2 = nil)
-      Git::DiffPathStatus.new(self, objectish, obj2)
+    # @param opts [Hash] Options to filter the diff.
+    # @option opts [String, Array<String>] :path_limiter Limit status to specified path(s).
+    # @option opts [String, Array<String>] :path (deprecated) Legacy alias for :path_limiter.
```
[the remark this answers is not in the document]
Comment (jcouball) [reply lib/git/base.rb]: Updated as requested.

File: lib/git/base.rb [hunk 9]
```diff
@@ -833,18 +833,32 @@ def merge_base(*)
...
 
     # Returns a Git::Diff::PathStatus object for accessing the name-status report.
     #
     # @param objectish [String] The first commit or object to compare. Defaults to 'HEAD'.
     # @param obj2 [String, nil] The second commit or object to compare.
-    # @return [Git::Diff::PathStatus]
-    def diff_path_status(objectish = 'HEAD', obj2 = nil)
-      Git::DiffPathStatus.new(self, objectish, obj2)
+    # @param opts [Hash] Options to filter the diff.
+    # @option opts [String, Array<String>] :path_limiter Limit status to specified path(s).
```
[the remark this answers is not in the document]
Comment (jcouball) [reply lib/git/base.rb]: I have updated this as requested

File: lib/git/base.rb [hunk 10]
```diff
@@ -833,18 +833,32 @@ def merge_base(*)
     #
     # @param objectish [String] The first commit or object to compare. Defaults to 'HEAD'.
     # @param obj2 [String, nil] The second commit or object to compare.
-    # @return [Git::Diff::Stats]
-    def diff_stats(objectish = 'HEAD', obj2 = nil)
-      Git::DiffStats.new(self, objectish, obj2)
+    # @param opts [Hash] Options to filter the diff.
+    # @option opts [String, Array<String>] :path_limiter Limit stats to specified path(s).
```
[the remark this answers is not in the document]
Comment (jcouball) [reply lib/git/base.rb]: Updated as requested.

File: lib/git/diff.rb [hunk 11]
```diff
@@ -18,8 +18,39 @@ def initialize(base, from = nil, to = nil)
...
+    # @example Limit diff to a single path
+    #   git.diff('HEAD~3', 'HEAD').path('lib/')
+    #
+    # @example Limit diff to multiple paths
+    #   git.diff('HEAD~3', 'HEAD').path('src/', 'docs/', 'README.md')
+    #
+    # @example Remove path filtering (show all files)
+    #   diff.path  # or diff.path(nil)
+    #
+    # @param paths [String, Pathname] one or more paths to filter the diff.
```
[the remark this answers is not in the document]
Comment (jcouball) [reply lib/git/diff.rb]: Updated as requested.
